src/nearest_search.py

```diff
@@ -0,0 +1,309 @@
+            # the publishing message.
+            for agent2 in data[self.num_agents1:self.num_agents]:
+                friend_position = agent2.pose.pose.position
+                distance = get_distance(agent_position, friend_position)
+                if distance > 0 and distance <= self.search_radius:
+                    # Calculate relative position
+                    rel_pos = deepcopy(agent2)
+                    rel_pos.pose.pose.position.x = friend_position.x - agent_position.x
+                    rel_pos.pose.pose.position.y = friend_position.y - agent_position.y
+                    nearest_agents.array.append(rel_pos)
+
+
+            for agent3 in data[0:self.num_agents1]:
+                friend_position = agent3.pose.pose.position
+                distance = get_distance(agent_position, friend_position)
+                if distance > 0 and distance <= self.search_radius:
+                    # Calculate relative position
+                    rel_pos = Pose()
+                    rel_pos.position.x = friend_position.x - agent_position.x
+                    rel_pos.position.y = friend_position.y - agent_position.y
+                    avoids.poses.append(rel_pos)
+
+            # Publish the relative positions message.
+            self.nearest[key].publish(nearest_agents)
+
+            # Initialize the wall and obstacle positions message.
+
+
+            # Positions of walls and obstacles are represented as value 100 in
+            # the list `self.map`. First, find the position of the observed
+            # agent in this list in form of an index pair. Then search the list
+            # in specified search radius and return actual positions of the
+            # walls and other obstacles.
+            or_col, or_row = self.pos_to_index(agent_position.x, agent_position.y)
+            # if key == 'sphero_0': print (or_col, or_row)
+            col_range = range(max(0, or_col - self.r), min(self.map_width, or_col + self.r + 1))
+            row_range = range(max(0, or_row - self.r), min(self.map_height, or_row + self.r + 1))
+            for row in row_range:
+                for col in col_range:
+                    # if key == 'sphero_0': print(self.map[row][col]/100, end=' ')
+                    # Check only elements within radius.
+                    # Search obstacles in a circle instead of a square.
+                    if (pow(row - or_row, 2) + pow(col - or_col, 2)) <= pow(self.r, 2):
+                        if self.map[row][col] == 100:
+                            x, y = self.index_to_pos(row, col)
+                            obst = Pose()
+                            obst.position.x = x - agent_position.x
+                            obst.position.y = y - agent_position.y
+                            avoids.poses.append(obst)
+                # if key == 'sphero_0': print('\n')
+
+            # Publish the wall and obstacle positions message.
+
+
+
+            self.avoid[key].publish(avoids)
```

Comment:
Ovdje ima jako puno duplog koda koji se ne potrebno izvršava dvaput. To bi sve trebalo svesti na jedan callback, parametarski odrediti početak i kraj liste agenata kroz koju se prolazi, možda to sve staviti u neke funkcije itd.